server/main.cpp

```diff
@@ -0,0 +1,178 @@
+#include "uvw.hpp"
+#include <memory>
+#include <iostream>
+#include <random>
+#include "connection.hpp"
+#include "packets.hpp"
+
+constexpr const char *addr = "127.0.0.1";
+constexpr const int port = 4242;
+
+void acceptClient(uvw::TCPHandle &srv);
+
+template<Packet T>
+void broadcast(uvw::Loop &loop, std::string &game_id, const T &packet);
+
+void handleLogin(uvw::TCPHandle &handle, PacketReader &reader);
+
+std::string genGameId() {
+    int len = 4;
+    static auto &chrs = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+    thread_local static std::mt19937 rg{std::random_device{}()};
+    thread_local static std::uniform_int_distribution<std::string::size_type> pick(0, sizeof(chrs) - 2);
+
+    std::string s;
+
+    s.reserve(len);
+
+    while (len--)
+        s += chrs[pick(rg)];
+
+    return s;
+}
+
+struct HandleData {
+    std::string nickname;
+    std::string game_id;
+    bool is_host;
+};
+
+std::vector<std::string> games;
+
+int main() {
+    auto loop = uvw::Loop::getDefault();
+    std::shared_ptr<uvw::TCPHandle> tcp = loop->resource<uvw::TCPHandle>();
+
+    tcp->on<uvw::ListenEvent>([](const uvw::ListenEvent &, uvw::TCPHandle &srv) {
+        acceptClient(srv);
+    });
+
+    tcp->bind(addr, port);
+    tcp->listen();
+    std::cout << "Server running at: " << addr << ":" << port << std::endl;
+    loop->run();
+}
+std::vector<char> recv_buffer{};
+
+void acceptClient(uvw::TCPHandle &srv) {
+    //Allocate new TCPHandle for a client
+    auto client = srv.loop().resource<uvw::TCPHandle>();
+
+    //Listeners
+    client->on<uvw::CloseEvent>([](const uvw::CloseEvent &, uvw::TCPHandle &client) {
+        std::cout << "[" << client.peer().ip << "]" << " disconnected " << std::endl;
+        client.close();
+    });
+    client->on<uvw::DataEvent>([&](const uvw::DataEvent &evt, uvw::TCPHandle &client) {
+        std::cout << "[" << client.peer().ip << "]" << " Received bytes: " << evt.length << std::endl;
+
+        std::copy_n(evt.data.get(), evt.length, std::back_inserter(recv_buffer));
+        std::cout << "buffer size: " << recv_buffer.size() << std::endl;
+        if(recv_buffer.size() < 4){
+            return;
+        }
+
+        auto reader = PacketReader(recv_buffer);
+        auto size = reader.readUInt();
+        if(recv_buffer.size() < size){
+            std::cout << "missing data size: " << size - recv_buffer.size() << std::endl;
```

Comment:
Throw zamiast wypisywać

Reply:
throw nie bo to wrecz normalny stan w ktorym nie dotarly wszystkie czesci pakietu, wypiske zakomentowalem